packages/eslint-plugin/tests/rules/key-spacing.test.ts

```diff
@@ -0,0 +1,428 @@
+  prop: {
+    abc: number;
+    a:   number;
+  },
+  abc:  string
+}
+      `,
+      output: `
+interface X {
+  a:    number;
+  prop: {
+    abc: number;
+    a:   number;
+  },
+  abc: string
+}
+      `,
+      options: [{ align: 'value' }],
+      errors: [{ messageId: 'extraValue' }],
+    },
+    {
+      code: `
+class X {
+  a:      number;
+  prop: {
+    abc: number;
+    a?: number;
+  };
+  abc: string;
+  x = 1;
+  d: number;
+  z:  number = 1;
+  ef: string;
+}
+      `,
+      output: `
+class X {
+  a:    number;
+  prop: {
+    abc: number;
+    a?:  number;
+  };
+  abc: string;
+  x = 1;
+  d:   number;
+  z:   number = 1;
+  ef:  string;
+}
+      `,
+      options: [{ align: 'value' }],
+      errors: [
+        { messageId: 'extraValue' },
+        { messageId: 'missingValue' },
+        { messageId: 'missingValue' },
+        { messageId: 'missingValue' },
+        { messageId: 'missingValue' },
+      ],
+    },
+    // align: colon
+    {
+      code: 'interface X {\n  a   : number;\n  abc: string\n};',
+      output: 'interface X {\n  a  : number;\n  abc: string\n};',
+      options: [{ align: 'colon' }],
+      errors: [{ messageId: 'extraKey' }],
+    },
+    // no align
+    {
+      code: 'interface X {\n  [x: number]:  string;\n}',
+      output: 'interface X {\n  [x: number]: string;\n}',
+      errors: [{ messageId: 'extraValue' }],
+    },
+    {
+      code: 'interface X {\n  [x: number]:string;\n}',
+      output: 'interface X {\n  [x: number]: string;\n}',
+      errors: [{ messageId: 'missingValue' }],
+    },
+    // singleLine / multiLine
+    {
+      code: 'interface X {\n  a:number;\n  abc:string\n};',
+      output: 'interface X {\n  a : number;\n  abc : string\n};',
+      options: [
+        {
+          singleLine: { beforeColon: false, afterColon: false },
+          multiLine: { beforeColon: true, afterColon: true },
+        },
+      ],
+      errors: [
+        { messageId: 'missingKey' },
+        { messageId: 'missingValue' },
+        { messageId: 'missingKey' },
+        { messageId: 'missingValue' },
+      ],
+    },
+    {
+      code: 'interface X { a : number; abc : string; };',
+      output: 'interface X { a:number; abc:string; };',
+      options: [
+        {
+          singleLine: { beforeColon: false, afterColon: false },
+          multiLine: { beforeColon: true, afterColon: true },
+        },
+      ],
+      errors: [
+        { messageId: 'extraKey' },
+        { messageId: 'extraValue' },
+        { messageId: 'extraKey' },
+        { messageId: 'extraValue' },
+      ],
+    },
+    {
+      code: 'interface X { a : number; abc : string; };',
+      output: 'interface X { a: number; abc: string; };',
+      options: [
+        {
+          singleLine: { beforeColon: false, afterColon: true },
+          multiLine: { beforeColon: true, afterColon: true },
+        },
+      ],
+      errors: [{ messageId: 'extraKey' }, { messageId: 'extraKey' }],
+    },
+  ],
+});
```

Comment:
Missing test coverage: wild and wacky types 😄. There should be test cases exercising...

* Mapped types
* Nested types (`outer: { middle: { inner: { /* ... */ } /* ... */ } }`)
* ...maybe other wild and wacky things, if you can think of them?

Reply:
There are already tests for mapped types (eg checks for `IndexSignature`), and nested types although only 1-deep.

Will try to make a wacky example though :)